prov/usnic/src/usdf_poll.c

```diff
@@ -0,0 +1,277 @@
+
+#include <fi.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+#include "usdf.h"
+#include "usdf_poll.h"
+
+static int usdf_poll_poll(struct fid_poll *pollset, void **context, int count)
+{
+	struct usdf_poll *_poll;
+	struct dlist_entry *item;
+	struct fid_list_entry *entry;
+	int ret = 0;
+	int copied = 0;
+	int pending;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset || !context) {
+		USDF_WARN_SYS(DOMAIN, "pollset and context can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	_poll = poll_ftou(pollset);
+
+	fastlock_acquire(&_poll->lock);
+
+	dlist_foreach(&_poll->list, item) {
+		entry = container_of(item, struct fid_list_entry, entry);
+		switch (entry->fid->fclass) {
+		case FI_CLASS_CQ:
+			pending = usdf_check_progress(entry->fid);
+			if ((pending == FI_SUCCESS) && (copied < count))
+				context[copied++] = entry->fid->context;
+			else if ((pending < 0) && (pending != -FI_EAGAIN))
+				ret = pending;
+			break;
+		default:
+			USDF_WARN_SYS(DOMAIN, "invalid object\n");
+			break;
+		}
+	}
+
+	fastlock_release(&_poll->lock);
+
+	if ((copied == count) || (copied > 0))
+		return copied;
+
+	return ret;
+}
+
+static int usdf_poll_add(struct fid_poll *pollset, struct fid *event_fid,
+		uint64_t flags)
+{
+	struct usdf_poll *_poll;
+	struct usdf_cq *_cq;
+	int ret;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset || !event_fid) {
+		USDF_WARN_SYS(DOMAIN, "pollset and event_fid can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	_poll = poll_ftou(pollset);
+
+	ret = fid_list_insert(&_poll->list, &_poll->lock, event_fid);
+	if (ret)
+		return ret;
+
+	switch (event_fid->fclass) {
+	case FI_CLASS_CQ:
+		_cq = cq_fidtou(event_fid);
+		ret = atomic_inc(&_cq->cq_refcnt);
+		USDF_DBG_SYS(DOMAIN,
+				"associated with CQ: [%p] with new refcnt: [%d]\n",
+				_cq, ret);
+		break;
+	default:
+		USDF_WARN_SYS(DOMAIN, "invalid fid class.\n");
+		return -FI_EINVAL;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int usdf_poll_del(struct fid_poll *pollset, struct fid *event_fid,
+		uint64_t flags)
+{
+	struct usdf_poll *_poll;
+	struct usdf_cq *_cq;
+	int ret;
+
+	if (!pollset || !event_fid) {
+		USDF_WARN_SYS(DOMAIN, "pollset and event_fid can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	_poll = poll_ftou(pollset);
+
+	switch (event_fid->fclass) {
+	case FI_CLASS_CQ:
+		_cq = cq_fidtou(event_fid);
+		ret = atomic_dec(&_cq->cq_refcnt);
+
+		USDF_DBG_SYS(DOMAIN,
+				"disassociating from CQ: [%p] with new refcnt: [%d]\n",
+				_cq, ret);
+		break;
+	default:
+		USDF_WARN_SYS(DOMAIN, "invalid fid class.\n");
+		return -FI_EINVAL;
+	}
+
+	fid_list_remove(&_poll->list, &_poll->lock, event_fid);
+
+	return FI_SUCCESS;
+}
+
+static int usdf_poll_close(struct fid *pollset)
+{
+	struct usdf_poll *_poll;
+	struct dlist_entry *item;
+	struct dlist_entry *head;
+	struct fid_list_entry *entry;
+	struct usdf_cq *_cq;
+	int ret;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset) {
+		USDF_WARN_SYS(DOMAIN, "pollset can't be NULL.\n");
```

Comment:
Nevermind this, I mixed up my enums and looked at the `FI_CLASS_` ones instead of the `FI_LOG_` ones.  There isn't any `FI_LOG_POLL` identifier.  Sorry for the noise.